examples/interpreter-demo/serialize.js

```diff
@@ -0,0 +1,411 @@
+        break;
+      default:
+        throw TypeError('Unknown type: ' + obj);
+    }
+    var props = Object.create(null);
+    var nonConfigurable = [];
+    var nonEnumerable = [];
+    var nonWritable = [];
+    var getter = [];
+    var setter = [];
+    var names = Object.getOwnPropertyNames(obj);
+    for (var j = 0; j < names.length; j++) {
+      var name = names[j];
+      if (jsonObj['type'] === 'Node' && name === 'loc') {
+        // Compactly serialize the location objects on a Node.
+        var loc = obj.loc;
+        var locText = '';
+        if (loc.start) {
+          locText += loc.start.line + ':' + loc.start.column;
+        } else {
+          locText += ':';
+        }
+        locText += '-';
+        if (loc.end) {
+          locText += loc.end.line + ':' + loc.end.column;
+        } else {
+          locText += ':';
+        }
+        if (loc.source !== undefined) {
+          locText += ' ' + encodeURI(loc.source);
+        }
+        jsonObj['loc'] = locText;
+      } else {
+        var descriptor = Object.getOwnPropertyDescriptor(obj, name);
+        props[name] = encodeValue(descriptor.value);
+        if (!descriptor.configurable) {
+          nonConfigurable.push(name);
+        }
+        if (!descriptor.enumerable) {
+          nonEnumerable.push(name);
+        }
+        if (!descriptor.writable) {
+          nonWritable.push(name);
+        }
+        if (descriptor.get) {
+          getter.push(name);
+        }
+        if (descriptor.set) {
+          setter.push(name);
+        }
+      }
+    }
+    if (names.length) {
+      jsonObj['props'] = props;
+    }
+    if (nonConfigurable.length) {
+      jsonObj['nonConfigurable'] = nonConfigurable;
+    }
+    if (nonEnumerable.length) {
+      jsonObj['nonEnumerable'] = nonEnumerable;
+    }
+    if (nonWritable.length) {
+      jsonObj['nonWritable'] = nonWritable;
+    }
+    if (getter.length) {
+      jsonObj['getter'] = getter;
+    }
+    if (setter.length) {
+      jsonObj['setter'] = setter;
+    }
+  }
+  return json;
+}
+
+// Recursively search the stack to find all non-primitives.
+function objectHunt_(node, objectList) {
+  if (node && (typeof node === 'object' || typeof node === 'function')) {
+    if (objectList.indexOf(node) !== -1) {
+      return;
+    }
+    objectList.push(node);
+    if (typeof node === 'object') {  // Recurse.
+      var isAcornNode = Object.getPrototypeOf(node) === NODE_CONSTRUCTOR.prototype;
+      var names = Object.getOwnPropertyNames(node);
+      for (var i = 0; i < names.length; i++) {
+        if (isAcornNode && names[i] === 'loc') {
+          continue;  // Skip over node locations, they are specially handled.
+        }
+        try {
+          objectHunt_(node[names[i]], objectList);
+        } catch (e) {
+          // Accessing some properties may trigger a placeholder getter.
+          // Squelch this error, but re-throw any others.
+          if (e.message !== 'Placeholder getter') {
```

Comment:
Which properties? You don't have a try/catch around the initial calls to `objectHunt_`. Why is that safe?

Reply:
Rewritten to make it clear that it's not the objectHunt_ call that might throw, but just the `node[names[i]]` property lookup.